[stdlib] Add support for the mapping protocol to Python interop

winding-lines opened the pull request:
Currently if you define a `__getitem__` method and export the struct to Python you can call the 
given method but you cannot use []. That's because the methods of the Mapping protocol need 
to be plumbed through special `tp_slots` in the type definition.

This PR proposes an approach for how this could be done. I am not familiar with current and future
directions of the Python Interop design. There are many of these type slots here:
https://github.com/python/cpython/blob/main/Include/typeslots.h. The design should accommodate
all those.

Feedback appreciated.

winding-lines committed:
Add support for the mapping protocol to Python interop

https://docs.python.org/3/c-api/typeobj.html#mapping-object-structures

JoeLoser commented on mojo/stdlib/std/python/bindings.mojo:
```diff
@@ -902,6 +1142,113 @@ struct PythonTypeBuilder(Copyable):
+        References:
+        - https://docs.python.org/3/c-api/typeobj.html#c.PyMappingMethods.mp_length .
+        """
+
```
```suggestion
            slot: The Python type slot to define, must be `mp_length`.
```

winding-lines replied:
done

JoeLoser commented on mojo/stdlib/std/python/bindings.mojo:
```diff
@@ -902,6 +1142,113 @@ struct PythonTypeBuilder(Copyable):
+        References:
+        - https://docs.python.org/3/c-api/typeobj.html#c.PyMappingMethods.mp_ass_subscript .
+        """
+
```
```suggestion
            slot: The Python type slot to define, must be `mp_setitem`.
```

winding-lines replied:
done

JoeLoser commented on mojo/stdlib/std/python/bindings.mojo:
```diff
@@ -902,6 +1142,113 @@ struct PythonTypeBuilder(Copyable):
+        """
+
+        self._insert_slot(
+            PyType_Slot.mp_subscript(_mp_subscript_wrapper[method])
```
```suggestion
            slot: The Python type slot to define, must be `mp_getitem`.
```

winding-lines replied:
done

JoeLoser commented on mojo/stdlib/std/python/bindings.mojo:
```diff
@@ -224,6 +225,245 @@ fn _tp_repr_wrapper[T: Representable](py_self: PyObjectPtr) -> PyObjectPtr:
+    Args:
+        py_self: Pointer to the Python object (the container).
+        key: Pointer to the Python object used as the subscript key.
+        value: Pointer to the value to set, or NULL for deletion.
```
The docstring mentions that `value` being NULL indicates deletion (`__delitem__`), but this code unconditionally creates a `PythonObject` from the borrowed pointer. If `value` is NULL (when Python calls `del obj[key]`), this could cause issues.

Consider either:
1. Adding a check for `value.is_null()` and handling deletion separately
2. Documenting that `__delitem__` is not supported by this wrapper
3. Providing a separate wrapper or signature for deletion support

winding-lines replied:
I used a Variant to support __delitem__

JoeLoser commented on mojo/integration-test/python-extension-modules/basic-columnar/mojo_module.mojo:
```diff
@@ -0,0 +1,184 @@
+        # For other comparisons, return False
+        return False
+
+
```
Returning `False` for unsupported comparison operations is incorrect behavior. Python's comparison protocol expects `NotImplemented` to be returned so the interpreter can try the reflected operation on the other operand.

For example, `a == b` where `a.__eq__(b)` returns `NotImplemented` will then try `b.__eq__(a)`.

Since users will likely use this as a template, consider either:
1. Implementing all comparison ops
2. Returning `NotImplemented` (may need support in the wrapper)
3. Adding a comment explaining this is a limitation of the example

winding-lines replied:
Fixed to support NotImplemented and use in the example. According to the docs exceptions are very cheap in Mojo so I went down that route. Happy to implement a different design if you prefer.

JoeLoser commented on mojo/integration-test/python-extension-modules/basic-columnar/mojo_module.mojo:
```diff
@@ -0,0 +1,184 @@
+from utils import IndexList
+from python import Python, PythonObject
+from python._cpython import PyObjectPtr, Py_LT, Py_GT
+from python.bindings import (
```
Nit: `PyObjectPtr` is imported but not used in this file.

winding-lines replied:
Removed, thanks!

JoeLoser commented on mojo/stdlib/std/python/_cpython.mojo:
```diff
@@ -407,11 +429,15 @@ struct PyType_Spec:
 
 
 # https://github.com/python/cpython/blob/main/Include/typeslots.h
+comptime Py_mp_setitem = 3
```
These constants are also defined as magic numbers in `PyTypeObjectSlot` in `bindings.mojo`. Consider having `PyTypeObjectSlot` reference these constants to avoid duplication:

```mojo
comptime mp_length = PyTypeObjectSlot(Py_mp_length)
```

soraros replied:
And maybe we should type them

soraros commented on mojo/stdlib/std/python/bindings.mojo:
```diff
@@ -902,6 +1142,113 @@ struct PythonTypeBuilder(Copyable):
+    fn def_method[
+        method: fn (PythonObject, PythonObject, Int) raises -> Bool,
+        slot: PyTypeObjectSlot,
+    ](mut self: Self,) -> ref [self] Self where slot.is_tp_richcompare():
```
Actually, I don't quite understand this API design. Why do don't we simply call the method `def_richcompare` and remove the type tag? Or if we want to keep the tag, we could make the "payload" a parameter, so the `where` clause is not necessary.

That said, `where` clause with a enum is a really interesting pattern.

winding-lines replied:
Agree, there are 2 design patterns for the API. Do we want one method name with type tags or do we want multiple and more specific method names. 

Like you said, the fact that the where clause worked was pretty cool and it's simpler to  have a small number of functions, so in this PR I went for that approach. Easy to change if the team prefers the second approach.

winding-lines committed:
Address feedback

Add support for __del_item__, ie NULL value in mp_ass_setitem.
Properly support not implemented for rich_compare including tests.
Cleanup docs.
Add typing around integer constats for type slots and rich compare ops.

winding-lines committed:
Port to mojo 20260131

winding-lines committed:
Port to mojo 26.2.0.dev2026030405
